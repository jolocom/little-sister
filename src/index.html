<!DOCTYPE html>

<html>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport"
          content="width=device-width, initial-scale=1, user-scalable=no">

    <title>Little Sister</title>

    <link href="https://fonts.googleapis.com/css?family=Roboto:100,400,300,600"
          rel="stylesheet"
          type="text/css" />
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons"
          rel="stylesheet">

    <style type="text/css">
      * {
        margin: 0;
        padding: 0;
      }
      html, body {
        height: 100%;
      }
      body {
        font-family: Roboto;
      }
      #app {
        width: 100%;
        height: 100%;
      }
    </style>

    <!-- <meta http-equiv="Content-Security-Policy" content="default-src * data: gap: https://ssl.gstatic.com 'unsafe-eval'; style-src * 'unsafe-inline'; media-src *; img-src * data:"> -->
    <meta name="format-detection" content="telephone=no">
    <meta name="msapplication-tap-highlight" content="no">
    <meta name="viewport" content="user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1, width=device-width">
  </head>
  <body>
    <div id="app"></div>
    <script src="https://solid.github.io/releases/rdflib.js/rdflib-0.7.0.min.js"></script>
    <script src="https://d3js.org/d3.v4.min.js"></script>
    <script src="js/bundle.js"></script>
  </body>
</html>
