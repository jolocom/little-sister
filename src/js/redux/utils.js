import _ from 'lodash'
import { bindActionCreators } from 'redux'
import { connect as reduxConnect } from 'react-redux'

/**
 * Abstraction on top of connect() from react-redux.
 *
 * In order to reduce boilerplate connecting components to Redux
 * and to prevent tightly coupling them to Redux itself, this decorator
 * is a drop-in replacement that allows you to specify the data and actions
 * you want to access declaritively.
 *
 * Suppose you have a Redux module named widget, exposing its data on
 * the lop-level global state, and with an action named foo, you could use
 * this like:
 *
 * connect({props: ['widget'], actions: ['widget:foo']})(
 *  props => <div onClick={props.foo}>
 *   {props.foo.data}
 *  </div>
 * )
 */
export function connect(params, wantedActions = []) {
  let wantedProps = params.props || params
  if (!wantedProps.map) {
    wantedProps = []
  }
  wantedActions = params.actions || wantedActions

  return reduxConnect(
    (state, props) => {
      if (typeof wantedProps !== 'function') {
        return _.fromPairs(wantedProps.map(prop => {
          const pair = [prop.split('.').slice(-1), state.getIn(prop)]
          if (pair[1].toJS) {
            pair[1] = pair[1].toJS()
          }
        }))
      } else {
        return wantedProps(state, props)
      }
    },
    (dispatch, props) => {
      if (typeof wantedActions !== 'function') {
        return bindActionCreators(_.fromPairs(wantedActions.map(id => {
          const [moduleName, actionName] = id.split(':')
          const module = require('redux/modules/' + moduleName)
          return [actionName, [actionName, module[actionName]]]
        })), dispatch)
      } else {
        return wantedActions(dispatch, props)
      }
    }
  )
}
