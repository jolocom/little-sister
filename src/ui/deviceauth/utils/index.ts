export const ERROR_TIMEOUT = 500
