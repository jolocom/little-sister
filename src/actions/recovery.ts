import { ThunkAction } from '../store'
//import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { navigationActions, genericActions } from './index'
import { routeList } from '../routeList'
import settingKeys from '../ui/settings/settingKeys'
import { removeNotification, scheduleNotification } from './notifications'
import { entropyToMnemonic, mnemonicToEntropy } from 'bip39'
 // TODO Import ^ from jolocom-lib
import useResetKeychainValues from 'src/ui/deviceauth/hooks/useResetKeychainValues'
import { PIN_SERVICE } from 'src/ui/deviceauth/utils/keychainConsts'
import { createWarningNotification } from 'src/lib/notifications'
import strings from 'src/locales/strings'
import I18n from 'src/locales/i18n'

export const showSeedPhrase = (): ThunkAction => async (
  dispatch,
  getState,
  agent,
) => {
  const encryptedSeed = await agent.storage
    .get.setting('encryptedSeed')

  if (!encryptedSeed) {
    throw new Error('Can not retrieve Seed from database')
  }

  const decrypted = await agent.idw.asymDecrypt(
    Buffer.from(encryptedSeed.b64Encoded, 'base64'),
    await agent.passwordStore.getPassword()
  )

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.SeedPhrase,
      params: { mnemonic: entropyToMnemonic(decrypted)},
    }),
  )
}

/**
 * called during PIN code restoration
 */
export const onRestoreAccess = (mnemonicInput: string[]): ThunkAction => async (
  dispatch,
  getState,
  agent,
) => {
  let recovered = false

  try {
    const recoveredEntropy = Buffer.from(
      mnemonicToEntropy(mnemonicInput.join(' ')),
      'hex'
    )
    if (agent.didMethod.recoverFromSeed) {
      const { identityWallet } = await agent.didMethod.recoverFromSeed(
        recoveredEntropy,
        await agent.passwordStore.getPassword()
      )
      recovered = identityWallet.did === agent.idw.did
    }
  } catch(e) {
    console.error('onRestoreAccess failed', e)
  }

  if (recovered) {
    const resetServiceValuesInKeychain = useResetKeychainValues(PIN_SERVICE)
    await resetServiceValuesInKeychain()
    dispatch(navigationActions.navigatorResetHome())

    // we re-lock the app, which will trigger the create pin screen
    return dispatch(genericActions.lockApp())
  } else {
    const notification = createWarningNotification({
      title: I18n.t(strings.AWKWARD),
      message: I18n.t(strings.IT_SEEMS_LIKE_WE_CANT_DO_THIS)
    })
    return dispatch(scheduleNotification(notification))
  }
}

export const setSeedPhraseSaved = (): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  // No delete call available yet, overwriting with empty object
  await backendMiddleware.storage.store.setting('encryptedSeed', {})
  await backendMiddleware.storage.store.setting(
    settingKeys.seedPhraseSaved,
    true,
  )

  const stickies = getState().notifications.queue.filter(n => !n.dismiss)

  stickies.map(sticky => dispatch(removeNotification(sticky)))

  return dispatch({
    type: 'SET_SEED_PHRASE_SAVED',
  })
}
