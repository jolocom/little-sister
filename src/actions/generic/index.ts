import { AppError, ErrorCode } from '../../lib/errors'
import I18n from '../../locales/i18n'
import { routeList } from '../../routeList'
import { navigationActions } from '../../actions'
import { ThunkAction } from '../../store'
import settingKeys from '../../ui/settings/settingKeys'

import * as Keychain from 'react-native-keychain'
// TODO use the settings items from storage
import AsyncStorage from '@react-native-community/async-storage'
// TODO don't depend on the crypto lib, perhaps use the rust crypto utils?
import crypto from 'crypto'

import { termsOfServiceDE } from 'src/ui/termsAndPrivacy/legalTexts'

import {
  AppWrapConfig,
  APPWRAP_UPDATE_CONFIG,
  APPWRAP_SHOW_LOADER,
  APPWRAP_REGISTER_CONFIG,
  APPWRAP_UNREGISTER_CONFIG,
  APPWRAP_SET_LOCKED,
  APPWRAP_SET_DISABLE_LOCK,
} from '../../reducers/generic'
import { AnyAction } from 'redux'
import { PIN_SERVICE } from 'src/ui/deviceauth/utils/keychainConsts'

// Default delay on the loading state value before it can switch back to 'false'
const DEFAULT_LOADING_LATCH_DELAY_MS = 300

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

const hashString = (text: string) => {
  return crypto
    .createHash('sha256')
    .update(text)
    .digest('hex')
}

const TERMS_OF_CONDITIONS_KEY = 'TERMS_OF_CONDITIONS'
export const checkTermsOfService = (
  route: routeList,
  onSubmit?: () => void,
): ThunkAction => async dispatch => {
  const storageHash = await AsyncStorage.getItem(TERMS_OF_CONDITIONS_KEY)
  const currentHash = hashString(termsOfServiceDE)
  const shouldShowTerms = storageHash !== currentHash

  if (!shouldShowTerms && onSubmit) onSubmit()

  if (shouldShowTerms) {
    return dispatch(
      navigationActions.navigate({
        routeName: routeList.TermsOfServiceConsent,
        params: { nextRoute: route, onSubmit },
      })
    )
  } else {
    return dispatch(
      navigationActions.navigate({
        routeName: route,
      })
    )
  }
}

export const storeTermsOfService = (
  route: routeList,
): ThunkAction => async dispatch => {
  const termsHash = hashString(termsOfServiceDE)
  await AsyncStorage.setItem(TERMS_OF_CONDITIONS_KEY, termsHash)

  dispatch(navigationActions.navigate({ routeName: route }))
}

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storage.store.setting(settingKeys.locale, {
    selected: locale
  })

  I18n.locale = locale
  dispatch({ type: 'SET_LOCALE', value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}

const setLoading = (value: boolean) => ({
  type: APPWRAP_SHOW_LOADER,
  value,
})

export const setLocked = (value: boolean) => ({
  type: APPWRAP_SET_LOCKED,
  value,
})

export const setDisableLock = (value: boolean) => ({
  type: APPWRAP_SET_DISABLE_LOCK,
  value,
})

export const lockApp = (
): ThunkAction => async (dispatch, getState) => {
  // if the lock is globally disabled, or there is no identity created/loaded
  // yet, then this action has no effect
  const disableLock = getState().generic.disableLock
  const hasIdentity = !!getState().account.did.did
  if (disableLock || !hasIdentity) return

  // otherwise we lock the app, using the appropriate lock screen
  const keychainPin = await Keychain.getGenericPassword({
    service: PIN_SERVICE,
  })
  const lockScreen = keychainPin ? routeList.Lock : routeList.RegisterPIN

  dispatch(setLocked(true))
  return dispatch(navigationActions.navigate({ routeName: lockScreen }))
}

export const unlockApp = (
  pin: string,
  resetNav?: boolean,
): ThunkAction => async dispatch => {
  const keychainPin = await Keychain.getGenericPassword({
    service: PIN_SERVICE,
  })

  if (!keychainPin || keychainPin.password !== pin) throw new Error('wrongPin')

  dispatch(setLocked(false))
  return resetNav
    ? dispatch(navigationActions.navigatorResetHome())
    : dispatch(navigationActions.navigateBack())
}

export const updateAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UPDATE_CONFIG,
  value,
})
export const registerAppWrapConfig = (value: AppWrapConfig & {}) => ({
  type: APPWRAP_REGISTER_CONFIG,
  value,
})
export const unregisterAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UNREGISTER_CONFIG,
  value,
})

export const showAppLoading = (
  shouldShow: boolean,
  latchDelay = DEFAULT_LOADING_LATCH_DELAY_MS,
): ThunkAction => async (dispatch, getState) => {
  const wasShowing = getState().generic.appWrapConfig.loading
  if (shouldShow && !wasShowing) {
    // for setting to "true" we do it immediately and set current state
    const ret = dispatch(setLoading(shouldShow))
    return new Promise<AnyAction>(resolve => {
      setTimeout(() => resolve(ret), latchDelay)
    })
  }

  if (wasShowing && !shouldShow) {
    // for setting to 'false', we delay by the latchDelay amount
    return dispatch(delayedSetLoading(shouldShow, latchDelay))
  }
}

let nextLoadingState = false
let timeout: number | undefined
const delayedSetLoading = (
  value: boolean,
  latchDelay: number,
): ThunkAction => async dispatch => {
  nextLoadingState = value
  if (timeout !== undefined) clearTimeout(timeout)
  timeout = setTimeout(() => {
    dispatch(setLoading(nextLoadingState))
    timeout = undefined
  }, latchDelay)
}
