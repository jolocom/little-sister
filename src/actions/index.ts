import { Interaction, FlowType } from '@jolocom/sdk'
import { ThunkAction } from 'src/store'

import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'
import * as recoveryActions from 'src/actions/recovery'
import * as notificationsActions from 'src/actions/notifications'
import strings from 'src/locales/strings'
import I18n from 'src/locales/i18n'
import { createInfoNotification } from '../lib/notifications'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
  recoveryActions,
  notificationsActions,
}

export const scheduleOfflineNotification = (
  message: string,
): ThunkAction => dispatch =>
  dispatch(
    notificationsActions.scheduleNotification(
      createInfoNotification({
        title: I18n.t(strings.NO_INTERNET_CONNECTION),
        message: I18n.t(message),
      }),
    ),
  )

export const interactionHandlers: {
  [flowType: string]: (interxn: Interaction) => ThunkAction
} = {
  [FlowType.Authentication]: ssoActions.consumeAuthenticationRequest,
  [FlowType.CredentialOffer]: ssoActions.consumeCredentialOfferRequest,
  [FlowType.CredentialShare]: ssoActions.consumeCredentialRequest,
  [FlowType.EstablishChannel]: ssoActions.consumeEstablishChannelRequest,
  [FlowType.Resolution]: ssoActions.consumeResolutionRequest,
}
