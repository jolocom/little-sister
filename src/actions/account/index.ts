import { groupBy, map, mergeRight, omit, uniq, zipWith } from 'ramda'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import { IdentitySummary, CredentialMetadataSummary } from '@jolocom/sdk'

import { navigationActions } from 'src/actions/'

import { routeList } from 'src/routeList'
import { CategorizedClaims, DecoratedClaims } from 'src/reducers/account'
import {
  getClaimMetadataByCredentialType,
  getCredentialUiCategory,
  getUiCredentialTypeByType,
} from 'src/lib/util'
import { ThunkAction } from 'src/store'
import { compose } from 'redux'
import { Not } from 'typeorm'
import { HAS_EXTERNAL_CREDENTIALS } from './actionTypes'
import { SDKError } from '@jolocom/sdk'
import { checkTermsOfService } from '../generic'
import { checkRecoverySetup } from '../notifications/checkRecoverySetup'

export const setDid = (did: string) => ({
  type: 'DID_SET',
  value: did,
})

export const setSelected = (claim: DecoratedClaims) => ({
  type: 'SET_SELECTED',
  selected: claim,
})

export const resetSelected = () => ({
  type: 'RESET_SELECTED',
})

export const handleClaimInput = (fieldValue: string, fieldName: string) => ({
  type: 'HANDLE_CLAIM_INPUT',
  fieldName,
  fieldValue,
})

export const checkIdentityExists: ThunkAction = async (
  dispatch,
  getState,
  agent,
) => {
  try {
    const identityWallet = await agent.loadIdentity()
    const userDid = identityWallet.identity.did
    dispatch(setDid(userDid))
    await dispatch(setClaimsForDid)
    await dispatch(checkRecoverySetup)
    return dispatch(checkTermsOfService(routeList.Home))
  } catch (err) {
    if (
      err.message === SDKError.codes.NoEntropy ||
      err.message === SDKError.codes.NoWallet
    ) {
      // No seed in database, user must register
      // But check if a registration was already in progress
      const isRegistering = getState().registration.loading.isRegistering

      const routeName = isRegistering
        ? routeList.RegistrationProgress
        : routeList.Landing

      return dispatch(navigationActions.navigate({ routeName }))
    }

    throw err
  }
}

export const openClaimDetails = (
  claim: DecoratedClaims,
): ThunkAction => dispatch => {
  dispatch(setSelected(claim))
  return dispatch(
    navigationActions.navigate({
      routeName: routeList.ClaimDetails,
    }),
  )
}

export const saveClaim: ThunkAction = async (
  dispatch,
  getState,
  agent,
) => {
  const { identityWallet, storage, passwordStore } = agent

  const did = getState().account.did.did
  const claimsItem = getState().account.claims.selected
  const password = await passwordStore.getPassword()

  const verifiableCredential = await identityWallet.create.signedCredential(
    {
      metadata: getClaimMetadataByCredentialType(claimsItem.credentialType),
      // the library acts directly on the object passed in, so a copy should be made first
      claim: { ...claimsItem.claimData },
      subject: did,
    },
    password,
  )

  if (claimsItem.id) {
    await storage.delete.verifiableCredential(claimsItem.id)
  }

  await storage.store.verifiableCredential(verifiableCredential)

  await dispatch(setClaimsForDid)

  return dispatch(navigationActions.navigatorResetHome())
}

export const hasExternalCredentials: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storage, identityWallet } = backendMiddleware
  // TODO FIXME
  // we only need a count, no need to actually load and deserialize
  // all of them
  const externalCredentials = await storage.get.verifiableCredential({
    issuer: Not(identityWallet.did),
  })

  return dispatch({
    type: HAS_EXTERNAL_CREDENTIALS,
    value: externalCredentials.length !== 0,
  })
}
export const setClaimsForDid: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storage } = backendMiddleware

  const verifiableCredentials: SignedCredential[] = await storage.get.verifiableCredential()

  const metadata = await Promise.all(
    verifiableCredentials.map(el => storage.get.credentialMetadata(el)),
  )

  const issuers = uniq(verifiableCredentials.map(cred => cred.issuer))

  const issuerMetadata = await Promise.all(
    issuers.map(storage.get.publicProfile),
  )

  const claims = prepareClaimsForState(
    verifiableCredentials,
    metadata,
    issuerMetadata,
  ) as CategorizedClaims

  return dispatch({
    type: 'SET_CLAIMS_FOR_DID',
    claims,
  })
}

export const prepareClaimsForState = (
  credentials: SignedCredential[],
  credentialMetadata: Array<CredentialMetadataSummary | {}>,
  issuerMetadata: IdentitySummary[],
) =>
  compose(
    groupBy(getCredentialUiCategory),
    zipWith(mergeRight, credentialMetadata),
    map(addIssuerInfo(issuerMetadata)),
    map(convertToDecoratedClaim),
  )(credentials)

export const addIssuerInfo = (issuerProfiles: IdentitySummary[]) => (
  claim: DecoratedClaims,
) => {
  if (!issuerProfiles || !issuerProfiles.length) {
    return claim
  }

  const issuer = issuerProfiles.find(el => el.did === claim.issuer.did)

  return issuer
    ? {
        ...claim,
        issuer,
      }
    : claim
}

/** @TODO Util, make subject mandatory (in lib) */
export const convertToDecoratedClaim = ({
  claim,
  type,
  issuer: issuerDid,
  id,
  expires,
}: SignedCredential): DecoratedClaims => ({
  credentialType: getUiCredentialTypeByType(type),
  issuer: {
    did: issuerDid,
  },
  claimData: omit(['id'], claim),
  id,
  subject: claim.id || 'Not found',
  expires: expires || undefined,
})
