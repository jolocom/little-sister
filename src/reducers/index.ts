import { combineReducers } from 'redux'
import { accountReducer, AccountState } from 'src/reducers/account/'
import {
  registrationReducer,
  RegistrationState,
} from 'src/reducers/registration/'
import { navigationReducer } from 'src/reducers/navigation/'
import { NavigationState } from 'react-navigation'
import { ssoReducer, SsoState } from 'src/reducers/sso/'

export const rootReducer = combineReducers({
  account: accountReducer,
  registration: registrationReducer,
  navigation: navigationReducer,
  sso: ssoReducer,
})

export interface RootState {
  readonly account: AccountState
  readonly registration: RegistrationState
  readonly navigation: NavigationState
  readonly sso: SsoState
}
