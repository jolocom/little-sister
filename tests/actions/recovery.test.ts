import data from './account/mockRegistrationData'
import { createMockStore, RecursivePartial } from 'tests/utils'
import { RootState } from 'src/reducers'

import { onRestoreAccess } from 'src/actions/recovery';
import { entropyToMnemonic} from 'bip39'


describe('Account recovery actions', () => {
  const initialState: RecursivePartial<RootState> = {
    account: {
      did: {
        did: 'some did'
      }
    },
    generic: {
      disableLock: false
    },
    notifications: {
      queue: []
    }
  }
  const { identityWallet, entropy, getPasswordResult } = data
  const mnemonic = entropyToMnemonic(Buffer.from(entropy, 'hex'))
  const mockMethod = {
    recoverFromSeed: jest.fn()
  }

  const agent = {
    identityWallet,
    passwordStore: {
      getPassword: () => getPasswordResult
    },
    didMethod: mockMethod
  }
  const mockStore = createMockStore(initialState, agent)
  const { resetGenericPassword, getGenericPassword } = require('react-native-keychain')

  beforeEach(() => {
    mockStore.reset()
    resetGenericPassword.mockReset()
  })

  it('should correctly restore access for forgotten pin', async () => {
    mockMethod.recoverFromSeed.mockImplementationOnce(() => {
      return {
        identityWallet: agent.identityWallet
      }
    })
    getGenericPassword.mockImplementationOnce(() => {
      // this is called with { service: PIN_SERVICE } and should return a PIN
      // but we expect the resetGenericPassword method to have been called, and
      // this should be empty
      return ''
    })
    await mockStore.dispatch(onRestoreAccess(mnemonic.split(' ')))

    expect(mockMethod.recoverFromSeed).toHaveBeenCalledWith(Buffer.from(entropy, 'hex'), getPasswordResult)
    expect(resetGenericPassword).toHaveBeenCalledTimes(1)
    expect(getGenericPassword).toHaveBeenCalledTimes(1)
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('should not restore access for incorrect mnemonic', async () => {
    const badEntropy = Buffer.from(entropy.replace('2', '4'), 'hex')
    const badMnemonic = entropyToMnemonic(badEntropy)
    mockMethod.recoverFromSeed.mockImplementationOnce(() => {
      return {
        identityWallet: {
          did: 'did:not:'
        }
      }
    })
    getGenericPassword.mockImplementationOnce(() => {
      // this is called with { service: PIN_SERVICE } and should return a PIN
      return 'TESTPIN'
    })
    await mockStore.dispatch(onRestoreAccess(badMnemonic.split(' ')))

    expect(mockMethod.recoverFromSeed).toHaveBeenCalledWith(badEntropy, getPasswordResult)
    expect(resetGenericPassword).toHaveBeenCalledTimes(0)
    expect(getGenericPassword).toHaveBeenCalledTimes(1)
    expect(mockStore.getActions()).toMatchSnapshot()
  })
})
