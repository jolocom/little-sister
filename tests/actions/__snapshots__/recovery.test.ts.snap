// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account recovery actions should correctly restore access for forgotten pin 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "routeName": "Home",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "key": null,
    "type": "Navigation/RESET",
  },
  Object {
    "type": "APPWRAP_SET_LOCKED",
    "value": true,
  },
  Object {
    "routeName": "RegisterPIN",
    "type": "Navigation/NAVIGATE",
  },
]
`;

exports[`Account recovery actions should not restore access for incorrect mnemonic 1`] = `
Array [
  Object {
    "notification": Object {
      "dismiss": Object {
        "timeout": 3000,
      },
      "id": 0,
      "message": "It seems like we can't do this",
      "title": "#awkward",
      "type": "warning",
    },
    "type": "SCHEDULE_NOTIFICATION",
  },
]
`;
